scipy/signal/_spectral_py.py

```diff
@@ -1874,44 +1896,53 @@ def detrend_func(d):
         # Wrap this function so that it receives a shape that it could
         # reasonably expect to receive.
         def detrend_func(d):
-            d = np.moveaxis(d, -1, axis)
+            d = xp.moveaxis(d, -1, axis)
             d = detrend(d)
-            return np.moveaxis(d, axis, -1)
+            return xp.moveaxis(d, axis, -1)
     else:
         detrend_func = detrend
 
-    if np.result_type(win, np.complex64) != outdtype:
-        win = win.astype(outdtype)
+    if xp.result_type(win, xp.complex64) != outdtype:
+        win = xp.astype(win, outdtype)
 
     if scaling == 'density':
-        scale = 1.0 / (fs * (win*win).sum())
+        scale = 1.0 / (fs * xp.sum(win*win))
     elif scaling == 'spectrum':
-        scale = 1.0 / win.sum()**2
+        scale = 1.0 / xp.sum(win) ** 2
     else:
         raise ValueError('Unknown scaling: %r' % scaling)
 
     if mode == 'stft':
-        scale = np.sqrt(scale)
+        scale = xp.sqrt(scale)
 
     if return_onesided:
-        if np.iscomplexobj(x):
+        is_complex = x.dtype in {xp.complex64, xp.complex128}
+
+        if is_complex:
             sides = 'twosided'
             warnings.warn('Input data is complex, switching to '
                           'return_onesided=False')
         else:
             sides = 'onesided'
             if not same_data:
-                if np.iscomplexobj(y):
+                if xp.iscomplexobj(y):
                     sides = 'twosided'
                     warnings.warn('Input data is complex, switching to '
                                   'return_onesided=False')
     else:
         sides = 'twosided'
 
     if sides == 'twosided':
-        freqs = sp_fft.fftfreq(nfft, 1/fs)
+        if hasattr(xp, 'fft'):
+            freqs = xp.fft.fftfreq(nfft, 1/fs)
+        else:
+            freqs = xp.asarray(np.fft.fftfreq(nfft, 1/fs))
+
     elif sides == 'onesided':
-        freqs = sp_fft.rfftfreq(nfft, 1/fs)
+        if hasattr(xp, 'fft'):
+            freqs = xp.fft.rfftfreq(nfft, 1/fs)
+        else:
+            freqs = xp.asarray(np.fft.rfftfreq(nfft, 1/fs))
```

Comment:
I think that `freqs = scipy.fft.fftfreq(..., xp=xp)` will work here, with negligible overhead.